cEP-0002.md

```diff
@@ -0,0 +1,402 @@
+An extra logger thread consumes time and resources of the system which we can
+avoid.
+
+The new proposal aims to eliminate the use of the queuing control signals.
+Utilising asynchronous I/O and process pool to generate results from  the
+bears. No longer will the local bears be dependent on global bears to finish.
+The logging will also be handled by the main thread itself instead of
+utilising an extra thread.
+
+# Proposed Change
+
+Here is the detailed implementation stepwise:
+
+1. A `ProcessPoolExecutor` is created with the maximum number of processes
+your CPU can run in parallel.
+2. We proceed with getting an asynchronous event loop. This asynchronous event
+loop creates `Concurrent.Futures` for tasks and schedules them to the process
+pool.
+3. Each task is given a callback when it has a result ready. This callback is responsible for processing the result accordingly.
+3. The `Bear` class will get two new methods. `execute_task()` and
+`generate_tasks()`.
+    * The `generate_tasks()` is resposible for generating the correct
+    arguments for the upcoming tasks.
+    * The `execute_task()` method is scheduled by the event loop to the
+    process pool using the arguments from the `generate_tasks()` method.
+3. Each bear will generate the following type of tasks:
+    * Global bears fire tasks that process all files.
+    * Local bears fire tasks for each individual file.
+    * Arbitrary tasks such as dependency bears or other tasks are also
+    possible.
+4. A bear that has dependencies will only be scheduled to the process pool once all its dependencies have been resolved. It's done in the following way:
+    * It's checked if the bear has some dependencies.
+    * If they haven't been resolved the bear is not scheduled.
+    * As soon as the dependencies are resolved for a bear it is removed from the dependency dictionary and scheduled to the `ProcessPoolExecutor`.
+5. A special purpose `Dependencies` class will be introduced to do the above
+mentioned task. This will be responsible for keeping track of dependencies.
+Whenever a bear completes it will be removed from the dependency class.
+6. When all the tasks are done the event loop is stopped and closed.
+7. You would not need to change any bears because coala will handle the
+backwards compatibility. There will be new versions of `LocalBear` and
+`GlobalBear` and will be called `FileWiseBear` and `ProjectWideBear`
+respectively.
+
+# The File Proxy
+
+With the new bear design we are also moving to a new wrapper object to wrap
+every file. This object will also contain some metadata such as the filename
+of the file. The file proxy is an object that contains properties that are
+cached and passed around processes.
+
+What it will store:
+    * The filename, this is the starting point for every other attribute.
+    * The content of the file as a string.
+    * The content of the file, line-splitted
```

Comment:
Um can't we just generate line-splitted contents from `contents`, when needed instead of storing it? I t seems like we are storing the same data twice. Correct me if I am missing something here :sweat_smile: